staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go

```diff
@@ -418,19 +480,128 @@ func (j *jitteredBackoffManagerImpl) Backoff() clock.Timer {
 // 3. a sleep truncated by the cap on duration has been completed.
 // In case (1) the returned error is what the condition function returned.
 // In all other cases, ErrWaitTimeout is returned.
+//
+// Since backoffs are often subject to cancellation, we recommend using
+// ExponentialBackoffWithContext and passing a context to the method.
 func ExponentialBackoff(backoff Backoff, condition ConditionFunc) error {
+	return ExponentialBackoffWithContext(context.Background(), backoff, condition.WithContext())
+}
+
+// ExponentialBackoffWithContext repeats a condition check with exponential backoff.
+// It immediately returns an error if the condition returns an error, the context is cancelled
+// or hits the deadline, or if the maximum attempts defined in backoff is exceeded (ErrWaitTimeout).
+// If an error is returned by the condition the backoff stops immediately. The condition will
+// never be invoked more than backoff.Steps times.
+func ExponentialBackoffWithContext(ctx context.Context, backoff Backoff, condition ConditionWithContextFunc) error {
+	// lazily initialize and reuse the timer
+	var after *time.Timer
+	defer func() {
+		if after != nil {
+			after.Stop()
+		}
+	}()
+
 	for backoff.Steps > 0 {
-		if ok, err := runConditionWithCrashProtection(condition); err != nil || ok {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+
+		if ok, err := runConditionWithCrashProtectionWithContext(ctx, condition); err != nil || ok {
 			return err
 		}
+
 		if backoff.Steps == 1 {
 			break
 		}
-		time.Sleep(backoff.Step())
+
+		waitBeforeRetry := backoff.Step()
+		if waitBeforeRetry == 0 {
+			continue
+		}
+		if after == nil {
+			after = timeNewTimer(waitBeforeRetry)
+		} else {
+			after.Reset(waitBeforeRetry)
+		}
+
+		select {
+		case <-after.C:
+		case <-ctx.Done():
+			return ctx.Err()
+		}
```

Comment:
it feels like this `select block` and the one at the beginning of the loop could be merged somehow ... but I can't say since I don't quite understand well the lazy initialization of the  timer, just thinking out loud

Reply:
Well spotted - there are some tricky behaviors in the various loop methods in this package, and this is one of them.  Specifically Backoff currently attempts to guarantee that once a context is cancelled you get at most one more invocation of the function, which because of the randomness of go select casing this construct doesn't guarantee, so you have to check twice.  I'm going to tweak the code slightly so this looks more like BackoffUntil - as a consequence of that we'll give that guarantee to all our methods.  During shutdown that sort of guarantee is useful to avoid excessive error logging that might confuse a user.

Reply:
See the more coherent (and now used across all three types of waits in this package) `loopConditionUntilContext` - has some comments inline.